File: platform/vcs-impl/backend/src/com/intellij/vcs/impl/backend/shelf/ShelveDeleteProvider.java
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.vcs.impl.backend.shelf;

import com.intellij.ide.DeleteProvider;
import com.intellij.openapi.actionSystem.ActionUpdateThread;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vcs.changes.shelf.ShelvedBinaryFile;
import com.intellij.openapi.vcs.changes.shelf.ShelvedChange;
import com.intellij.openapi.vcs.changes.shelf.ShelvedChangeList;
import com.intellij.openapi.vcs.changes.shelf.ShelvedChangesViewManager;
import com.intellij.util.ui.tree.TreeUtil;
import org.jetbrains.annotations.NotNull;

import java.util.List;

final class ShelveDeleteProvider implements DeleteProvider {
  private final @NotNull Project myProject;
  private final @NotNull ShelfTree myTree;

  private ShelveDeleteProvider(@NotNull Project project, @NotNull ShelfTree tree) {
    myProject = project;
    myTree = tree;
  }

  @Override
  public @NotNull ActionUpdateThread getActionUpdateThread() {
    return ActionUpdateThread.BGT;
  }

  @Override
  public void deleteElement(@NotNull DataContext dataContext) {
    List<ShelvedChangeList> shelvedListsToDelete = TreeUtil.collectSelectedObjectsOfType(myTree, ShelvedChangeList.class);
    List<ShelvedChangeList> shelvedListsFromChanges = ShelvedChangesViewManager.getShelvedLists(dataContext);
    List<ShelvedChange> selectedChanges = ShelvedChangesViewManager.getShelveChanges(dataContext);
    List<ShelvedBinaryFile> selectedBinaryChanges = ShelvedChangesViewManager.getBinaryShelveChanges(dataContext);

    ShelvedChangesViewManager.deleteShelves(myProject, shelvedListsToDelete, shelvedListsFromChanges, selectedChanges,
                                            selectedBinaryChanges);
  }

  @Override
  public boolean canDeleteElement(@NotNull DataContext dataContext) {
    return !ShelvedChangesViewManager.getShelvedLists(dataContext).isEmpty();
  }
}

File: platform/vcs-impl/backend/src/com/intellij/vcs/impl/backend/shelf/ShelveDeleteProvider.kt
// Copyright 2000-2024 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.vcs.impl.backend.shelf;

import com.intellij.ide.DeleteProvider
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.project.Project
import com.intellij.openapi.vcs.changes.shelf.ShelvedChangeList
import com.intellij.openapi.vcs.changes.shelf.ShelvedChangesViewManager
import com.intellij.util.ui.tree.TreeUtil

internal class ShelveDeleteProvider private constructor(project: Project, tree: ShelfTree) : DeleteProvider {
  private val myProject: Project
  private val myTree: ShelfTree

  init {
    myProject = project
    myTree = tree
  }

  override fun getActionUpdateThread(): ActionUpdateThread {
    return ActionUpdateThread.BGT
  }

  override fun deleteElement(dataContext: DataContext) {
    val shelvedListsToDelete = TreeUtil.collectSelectedObjectsOfType<ShelvedChangeList?>(myTree, ShelvedChangeList::class.java)
    val shelvedListsFromChanges = ShelvedChangesViewManager.getShelvedLists(dataContext)
    val selectedChanges = ShelvedChangesViewManager.getShelveChanges(dataContext)
    val selectedBinaryChanges = ShelvedChangesViewManager.getBinaryShelveChanges(dataContext)

    ShelvedChangesViewManager.deleteShelves(myProject, shelvedListsToDelete, shelvedListsFromChanges, selectedChanges,
                                            selectedBinaryChanges)
  }

  override fun canDeleteElement(dataContext: DataContext): Boolean {
    return !ShelvedChangesViewManager.getShelvedLists(dataContext).isEmpty()
  }
}
